XTS-16 aop解析时，应该考虑一个配置类对应一个目标类中的多个方法的多个配置的case

snowflake-Zhao opened the pull request:
…b7b0794f?

#XTS-16 aop解析时，应该考虑一个配置类对应一个目标类中的多个方法的多个配置的case

解析aop时候，应该以method为最小抽象单元
动态代理时，应该以class为最小抽象单元

snowflake-Zhao committed:
https://zrzr20220926114535774.pingcode.com/pjm/items/65c42dc204ae6f86b7b0794f?
#XTS-16 aop解析时，应该考虑一个配置类对应一个目标类中的多个方法的多个配置的case

解析aop时候，应该以method为最小抽象单元
动态代理时，应该以class为最小抽象单元

GUBIY commented on sys_design_proj/spring_engine/src/sdk/SpringApplication.java:
```diff
@@ -50,41 +50,40 @@ private void aop() throws ClassNotFoundException, IllegalAccessException, Instan
-      final Method afterMethod = aopConf.afterMethod;
-      final String targetMethod = aopConf.targetMethod;
+      Map<String, AopMethodConf> methodConfMap = aopConf.methodConfMap;
       Object aspectObj = Class.forName(aopConf.aspect).newInstance();
```
aspect对象这时候在initBeanContainer已经初始化过了，不需要再new一个新的对象吧？

snowflake-Zhao replied:
你要生成动态代理，肯定是要根据现有的对象生成代理对象蛤

snowflake-Zhao replied:
修了
https://github.com/GUBIY/SystemDesign/pull/8/commits/4592612e1c4eb377d8abc20342672645c8732802

GUBIY commented on sys_design_proj/spring_engine/src/sdk/SpringApplication.java:
```diff
@@ -93,12 +92,25 @@ private void aop() throws ClassNotFoundException, IllegalAccessException, Instan
           @Override
           public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
             String methodName = method.getName();
```
如果是不同的类里但是同名的方法，也会触发到before和after方法吗？

snowflake-Zhao replied:
会的。
因为我是以类为单元解析AOP的，也就是说一个AOP的实体类的对象有针对该类的所有方法的aop配置。
同名方法不同类会生成两个动态代理对象，会执行相同的方法判断语句，因此会执行

GUBIY commented on sys_design_proj/spring_engine/src/sdk/SpringApplication.java:
```diff
@@ -93,12 +92,25 @@ private void aop() throws ClassNotFoundException, IllegalAccessException, Instan
             String methodName = method.getName();
-            if (beforeMethod != null && methodName.equals(targetMethod)) {
-              beforeMethod.invoke(aspectObj, method, args);
+            for (String aopMethod : methodConfMap.keySet()) {
```
这个for循环遍历似乎没必要吧，methodConfMap.get(targetMethod)得到AopMethodConf，然后就直接可以判beforeMethod和afterMethod是不是为空了吧？

snowflake-Zhao replied:
一个类针对不同的方法可能会有多种before和after的规则，因为动态代理对象是依赖类动态生成的，所以要把类中所有的针对不同方法的aop规则，做处理一并写入
![Snipaste_2024-02-17_15-29-49](https://github.com/GUBIY/SystemDesign/assets/43769361/fd35ebf0-d310-44fe-8a15-62c1fe1d847d)
动态代理对象里面。
就像这个aop配置类对IGreetingServiceImpl类下面的farewell和greet的不同的行为

snowflake-Zhao committed:
https://zrzr20220926114535774.pingcode.com/pjm/items/65c42dc204ae6f86b7b0794f?
#XTS-16 aop解析时，应该考虑一个配置类对应一个目标类中的多个方法的多个配置的case

aspect 配置类应该也统一由bean容器管理，因此统一从bean copntainer中取出

GUBIY commented on sys_design_proj/spring_engine/src/sdk/SpringApplication.java:
```diff
@@ -50,41 +50,40 @@ private void aop() throws ClassNotFoundException, IllegalAccessException, Instan
               continue;
             }
-            classAopConfMap.get(aopConfigClass.getCanonicalName()).afterMethod = method;
+            classAopConfMap.get(aopConfigClass.getCanonicalName()).putAfterMethod(targetMethod, method);
```
我之前关于不同类的同名方法困惑，现在看起来是在里的问题， 即line 64 和line 78 。

AOP Target value 完整的定义是包含两部分的, 一个是target class , 一个是 target method. 这就意味着一个aopConfigClass, 不止可能包含多个target method，也可能包含多个target class。

否则如果有这么两个不同的aop方法@Before(value = "edu.ecnu.di.IGreetingServiceImpl1.greet") 和 @Before(value = "edu.ecnu.di.IGreetingServiceImpl2.greet")，就永远只能调用到IGreetingServiceImpl1  而调用不到IGreetingServiceImpl2 这个类

snowflake-Zhao replied:
爸爸我错了，我脑残自己写错了，key应该用targetclass，你看下新的commits -ref
https://github.com/GUBIY/SystemDesign/pull/8/commits/703149031c1efd479fe2066cefd9c3681d761aaa

GUBIY replied:
就很完美了

snowflake-Zhao committed:
#XTS-16 aop解析时，应该考虑一个配置类对应一个目标类中的多个方法的多个配置的case
我是脑残，写错了，应该是用targetclass用做key来进行aop配置中对method进行class级别的隔离
